### DISCUSSION TITLE
Physical Input

### DISCUSSION THREAD

**Discussion by nebhead** [nebhead/PiFire#18] [Q&A] [answered]
Just curious if anyone here would be interested in having any physical input buttons / selectors for PiFire?  Having a web interface is nice, but sometimes having physical buttons is nice too.  No promises, but I was thinking about adding a physical button interface module to PiFire.  

So my question to the "community" here is, if we added input, what kind of input would you like to see?  If just buttons, how many buttons?  What would they do?  Rotary encoder maybe?  Selector switch?

**Comment by faisalhusain007** [comment, marked as answer]
Yes, that would be awesome, I was thinking either a 3 button up down for temperature enter that controls start / stop. Prefer that to encoder.

**Reply by nebhead** [reply]
Excellent!  I'll get started on something like this.  I'm thinking three small tactile push button switches then, which can be mounted into your 3D printed enclosure.  I like your idea of up/down/enter.  I think that would be perfect for simplicity and for convenience.

**Reply by Aviator85** [reply]
![IMG_7465](https://user-images.githubusercontent.com/37892009/118426881-b0f08d00-b691-11eb-80b3-1134fbe144b4.JPG)
I was taking a nap earlier and actually came to my mind how an rotary encoder might be a slick route then got on here and seen that was already kinda snuffed out ☹️  I put one on my moode audio setup in the garage and yeah its not as precise but to me its a secondary means so if it takes a few extra turns back and forth I'd be okay with that.

**Reply by nebhead** [reply]
Hey @Aviator85 - I actually did order some rotary encoders from Amazon and played around with them for this project.  Ultimately, I decided that the 3-button setup was superior.  There are a couple good reasons for this: 

1. The rotary encoder implementation on the Pi, generally requires the use of interrupt callbacks.  My initial design for input was using callbacks, but I ran into serious issues with the interrupts spuriously firing even when there was no real physical inputs happening.  Another side effect of these interrupts firing off was that it sometimes happened while the display was being accessed / which caused the display to go into unknown states.  Unfortunately, I didn't want to go too far down the rabbit hole of chasing spurious interrupts and state switching.  
2. Per reason #1, I switched back to using button polling.  With button polling implemented, the performance of the rotary encoder suffers too much.  Lots of missed movements and button presses - which is just a bad user experience.  Ultimately, I found that the 3 button setup is actually really easy to use and might even be more convenient than the rotary encoder.  

You are welcome to take a look at the code in control.py (and your display module) and see if you want to give it a shot.  It's easy enough to create a new display module with rotary encoder support to try it out.

**Comment by nebhead** [comment]
Question to this group, maybe a call for assistance. 

I'm experiencing a lot of issues with button input and spurious/ghost button presses.  I've tried several different combinations of pull-ups/pulldowns with tactile switches, changing resistor sizes.  Nothing seems to work for me.  

Here's the schematic example of the button input with pullups.  I've tried changing the 10k ohm resistors with 4.7kohm resistors (to pull the pin up more.  Regardless, seems to be getting random ghost inputs.  

![PiFire-PiSide-ButtonInput](https://user-images.githubusercontent.com/19876886/110053453-4bd20000-7d0e-11eb-846d-9c75d67bfd7b.jpg)

The relevant code for setting up the GPIO's is here: 
```
		GPIO.setmode(GPIO.BCM)
		GPIO.setup(self.up, GPIO.IN, pull_up_down=GPIO.PUD_UP)
		GPIO.setup(self.down, GPIO.IN, pull_up_down=GPIO.PUD_UP)
		GPIO.setup(self.enter, GPIO.IN, pull_up_down=GPIO.PUD_UP)

		GPIO.add_event_detect(self.up, GPIO.FALLING, callback=self.UpCallback, bouncetime=300)  
		GPIO.add_event_detect(self.down, GPIO.FALLING, callback=self.DownCallback, bouncetime=300) 
		GPIO.add_event_detect(self.enter, GPIO.FALLING, callback=self.EnterCallback, bouncetime=300) 
```

I've tried to remove the internal Pull-Up definitions, I've tried changing the Falling Edge to Rising Edge.  I've increased the bounce time to 600.  Nothing seems to affect the issue.  

I'm thinking I may have to remove the event detect method and go to a polling method (which is not ideal for the design of PiFire).  

Thanks!

**Comment by faisalhusain007** [comment]
I had mine wired up slightly differently, there is no connection to 3.3v
everything else is the same, except for the pullups I didn't use them on
the inputs. I don't get any ghosting. I did however use a small Ferrite
Sleeve on the wires coming off the pcb. When I was testing without the
buttons connected but with the wires plugged in the PCB I would get some
ghosting. After installing the sleeve the problem seemed to be resolved. I
haven't tested the buttons on the new 2.8" screen but my wires are still on
the PCB (buttons, not connected), no ghosting experienced yet and I am
using display_ili9341b.

**Reply by nebhead** [reply]
> I had mine wired up slightly differently, there is no connection to 3.3v everything else is the same, except for the pullups I didn't use them on the inputs. I don't get any ghosting. I did however use a small Ferrite Sleeve on the wires coming off the pcb. When I was testing without the buttons connected but with the wires plugged in the PCB I would get some ghosting. After installing the sleeve the problem seemed to be resolved. I haven't tested the buttons on the new 2.8" screen but my wires are still on the PCB (buttons, not connected), no ghosting experienced yet and I am using display_ili9341b.

Interesting, I'll have to try the ferrite.  Definitely something with the length of the wires possibly causing some noise; antenna effect.  I think my power supply might be noisy as well (well maybe my mains power since this house was built in 1974 with aluminum wiring).  

I've tried many combinations of pull-ups and pull-downs on my breadboard, as well as adding capacitors as suggested by some online.  So far, nothing has worked.  It seems that these spurious inputs are pretty common when using the event_detect due to the way it detects edges internally.  

With all that said, I'm going to mess around with the software side to implement something a little less elegant, polling the inputs every so often.  This might actually be the best solution, given that we may be going with a touch screen input for future revisions - this will make it easier to get input from the touchscreen (if we ever figure that out).  :)

**Comment by faisalhusain007** [comment]
To be honest, On my latest panel I did away with the switch cutouts as I
was going to move forward with the touch screen, I did leave the wires
connected to the PCB, I haven't had any ghosting yet. However if I touch
the ends of the wires with my hands I can get the menu on and flick
through the options without problems.

**Reply by nebhead** [reply]
> To be honest, On my latest panel I did away with the switch cutouts as I was going to move forward with the touch screen, I did leave the wires connected to the PCB, I haven't had any ghosting yet. However if I touch the ends of the wires with my hands I can get the menu on and flick through the options without problems.

Yeah, I decided to rearchitect things so that we are polling the GPIO's instead of using the interrupts given the instability.  I actually really like the change because it allows us to easily add touch screen input (if/when we figure that out).  Also it allows a long press to cycle menu options instead of multiple pushes.  Responsiveness needs to be tuned, but basically it is working well without any ghosting issues.  I'll upload the code later this week once I get the circuit soldered up and tested.  

On the touch screen, do you happen to know what controller is on your display?  Is it the XPT2046?  I've not had a ton of luck finding something that will work alongside the existing libraries.  Might resort to going to the adafruit libraries.

**Comment by nebhead** [comment]
Ok, got it up and running and I think I like the button input with the new architecture.  As a side note, I think this also resolves an issue where callback interrupts were causing the screen to corrupt sometimes.  This is no longer an issue because all screen routines are allowed to run uninterrupted.  I also changed the circuitry to pull downs.  

![PiFire-Button-Input](https://user-images.githubusercontent.com/19876886/110739114-63652900-81e5-11eb-983e-2c818d4ba26b.jpg)

Here's a short demo with my current 3D printed panel.  

https://user-images.githubusercontent.com/19876886/110740766-71687900-81e8-11eb-9d14-469b5736b6ae.mp4

**Comment by faisalhusain007** [comment]
That looks good, the case looks sturdy and a good fit. The buttons look
better than the ones I printed up. I have had to travel back to the UK to
attend to some personal family business, but I did pack my pifire with the
touch panel belive it or not. Currently I have to serve my 10 day self
quarantine so got some time will unpack and take a look. But I got no grill
here unfortunately. And it's way too cold anyways.

**Reply by nebhead** [reply]
Thanks!  I'm really having fun with OnShape - it can be frustrating at times, but it's also much more powerful than TinkerCad and allows me to do things that would have taken hours there.  The buttons have a little bit of slop in them - meaning they are a little loose - but honestly it's probably as close as I'm going to get with my current skill level.  

And regarding your travel - wow!  I guess if I was going to be stuck in quarantine for 10 days I might also bring this along to play with. :)  Maybe you can help me look into the touch screen part, as I'd like to also replace the buttons with touch eventually (I'd like to order the bigger 3.2" touch display).  But as it stands, I almost feel like we will need to develop our own touch library for this.  

Stay safe!  Enjoy your trip.

**Comment by cptcave** [comment]
Has there been any movement on the touch screen capability here?  As I order my parts I'm wondering if I should forgo the buttons completely and just go with the touch screen.  Also, in your build manifest you mentioned a specific ILI9341 screen.  I found a different ILI9341 with good ratings but $10 cheaper from HiLetgo on Amazon.  Would that be compatible? It looks identical. 

Also, you've noted here that with the buttons 3 4.7 kohm resistors for the common connection to the gpio on the rpi.  Those aren't in the manifest, is this a change to that or is the manifest more up to date? I'm sorry, my knowledge of electrical engineering is limited to soldering.  Component configurations and ratings are all magic to me.

**Reply by nebhead** [reply]
No movement on the touch screen part.  It's going to require some re-architecture of the control.py script, I think.  Also, there are limited touch screen python libraries for the raspberry pi and might require creating or hacking one up from the ground up.  It would be cool, but it's something I've put on the back burner for now.  I rationalize this even more when I'm starting up the grill with messy hands, it's nice not having to touch the screen.  :)  

As for the ILI9341 - Just make sure you are getting the right screen dimensions for your build.  I'm using 320x240 in my build (and the module ili9341b is using this).  Generally, the HiLetGo stuff is fine - just tends to be re-branded generic.  

As for the resistors, just follow the schematic for button input here: https://nebhead.github.io/PiFire/.  You should be able to use anywhere from 4.7k to 10k or whatever you have laying around.  These are there to prevent shorting to ground.

**Reply by weberbox** [reply]
I read this comment and saw that it was suggested to use the pinout at https://nebhead.github.io/PiFire/  which is pull ups but the code is looking for pull downs. I wired up my display and it went crazy thinking I was pressing every button. I had to modify the display_ili9341b.py to look for pull ups with a 0 instead of 1. I added defines in mine to make the switch easier. 

![image](https://user-images.githubusercontent.com/313328/125404817-f2d54100-e36b-11eb-8fcd-b26fdc8ed0f4.png)

![image](https://user-images.githubusercontent.com/313328/125405028-21531c00-e36c-11eb-95cb-a9b6cf47c5bb.png)
![image](https://user-images.githubusercontent.com/313328/125404910-084a6b00-e36c-11eb-8274-ef1cec202d52.png)

**Reply by nebhead** [reply]
@zippy-man - Thanks for catching another dumb mistake of mine!  I'd be happy to update with a pull request if you want.

**Reply by weberbox** [reply]
@nebhead I saw commit https://github.com/nebhead/PiFire/commit/9d8e76e4181377be039ee6197119d56addf2a0b7 and how you added the HIGH and LOW relays. Should the buttons also be added this way instead of in each display class that uses buttons?

**Reply by nebhead** [reply]
Oh, I think it will have to be modified in each of the button display classes anyway, so maybe we just stick with your solution.  In each of the display_xxxxb.py files (files with a b), just basically replace the event detect with the above.  For some of us (me) with the reverse logic, I'll just have to remember to update this file and flip the bit whenever I do an update.  

Thanks, 
Ben

**Reply by weberbox** [reply]
I was already working on adding it as a setting in my button-levels branch, I can send that pull request if you want to check it out. If you want to just go with the Defines that is ok. I can make those changes and do a pull request for that. Just let me know

**Reply by nebhead** [reply]
Yeah, if you want to just sent a pull request so I can check out the changes, that would be great.  I think the defines would be just fine, but either way would be fine by me.

**Reply by weberbox** [reply]
OK I sent them both over. Use one or the other or neither your choice.  I am not amazing with python so please check. They both work for me so it should be ok.

**Reply by nebhead** [reply]
Thank you!  I'll take a look - if one of them works out, I'll approve and you'll be officially a contributor. :)   (BTW - I'm not great with Python either, but I'm learning)

**Reply by dgshue** [reply]
Anyone made any progress on the touch screen? I hadn't thought about the practicality of it (with greasy fingers) but I've got a display here just begging to be touched.  Thought I might attempt to tackle it if no one else has made any movement.

**Reply by nebhead** [reply]
Yeah, haven't made any movement on this one yet. :)  I kind of stalled out when I couldn't find suitable off-the-shelf Python touch libraries for the display that I have.  But I would encourage you to give it a try and see what you can come up with.

**Reply by weberbox** [reply]
This might be a good starting point. I have not really checked it out much but it should be for XPT2046 touch
https://github.com/MatthewLowden/RPi-XPT2046-Touchscreen-Python

I am not sure all the ILI9341 come with the same chip but the one below is what I bought

https://www.amazon.com/WayinTop-ILI9341-Display-240X320-Resistor/dp/B087C3PP9G/ref=mp_s_a_1_3?dchild=1&keywords=ili9341&qid=1628877548&sr=8-3

**Reply by nebhead** [reply]
Yep, I have an ILI9341 with XPT2046 as well, although that's not what I have on my grill at the moment.  Would be cool to get something like that to work.  If someone wants to experiment.  Since the display itself is being accessed via SPI, I wasn't sure if these types of bit-banged libraries would interfere with those.  It's worth a shot though.

**Comment by gsherwin74** [comment]
I use a Windows 10 tablet mounted on the garage wall and have Chrome
running fullscreen.  Just plug the smoker in, wait a few minutes for
everything to boot up and come online and then control it from the tablet.

**Reply by nebhead** [reply]
Well, that's a slick sounding setup!  You know, I have a few old android devices just laying around and collecting dust.  Would be perfect for this kind of usage.

**Reply by weberbox** [reply]
I saw this and immediately thought of some old tablets I have. Perfect use case 

It would be great to have a native Android app. I have pondered starting a project on that
